Finished work to allow Crucible.NPC to support custom customers and custom traders

AndrewFahlgren opened the pull request:
- Fixed 1.0 issue with custom icons
- Expanded existing code for Crucible.NPC customers
- Added traders to Crucible.NPC
- Added some basic skeleton work for Crucible.NPC factions

AndrewFahlgren committed:
Added some initial classes to specify custom traders, customers, and factions

AndrewFahlgren committed:
Finished basic trader configuration pieces

AndrewFahlgren committed:
Finished custom customer logic

AndrewFahlgren committed:
Commented out unfinished code

AndrewFahlgren committed:
Added basic dialogue system skeleton
Removed option to have NPCs come on a certain day since Potion Craft no longer supports it

AndrewFahlgren committed:
removed redundant cast

AndrewFahlgren committed:
Finished first round of NPC support

AndrewFahlgren committed:
Fixed several issues with custom customers
Refactored quest configuration

AndrewFahlgren committed:
Fixed remaining customer dialogue issues

AndrewFahlgren committed:
Fixed bugs with custom traders
Modified dialogue system to be more forgiving to modding errors
Modified appearance system to allow for beards
Fixed issue where some appearance parts were not being copied to the custom npc

AndrewFahlgren committed:
removed debug code for beards

AndrewFahlgren committed:
removed commented out code which is no longer relevant

AndrewFahlgren commented on Crucible.NPCs/CrucibleNpcAppearanceConfig.cs:
```diff
@@ -64,23 +64,9 @@ public class CrucibleNpcAppearanceConfig
         /// <param name="npc">The NPC to apply the configuration to.</param>
         public void ApplyAppearance(CrucibleNpcTemplate npc)
         {
-            npc.Appearance.Clear();
```
Removed this clear to allow for people to copy most of an NPC while only specifying appearance elements they want to change.

AndrewFahlgren commented on Crucible.NPCs/Quest/CrucibleNPCDialogueQuestConfig.cs:
```diff
@@ -14,38 +14,29 @@
+        /// <summary>
+        /// Gets or sets the starting dialogue node.
+        /// </summary>
+        public CrucibleDialogueData.CrucibleDialogueNode Dialogue { get; set; }
```
I am not sure if this is within design guidelines but I went ahead and used some GameAPI classes in the config for dialogue. The dialogue system I designed uses a node concept which didn't translate well to traditional config classes. To remedy this the best solution may be to move the `CrucibleDialogueData.CrucibleDialogueNode` and `CrucibleDialogueData.CrucibleAnswerNode` classes into Crucible.NPC and rename them to look like config classes.

AndrewFahlgren commented on Crucible.NPCs/Quest/CrucibleQuestRequirementConfig.cs:
```diff
@@ -0,0 +1,49 @@
+    [DuckTypeCandidate(typeof(CrucibleNoIngredientQuestRequirementConfig))]
+    [DuckTypeCandidate(typeof(CrucibleStrongPotionQuestRequirementConfig))]
+    [DuckTypeCandidate(typeof(CrucibleWeakPotionQuestRequirementConfig))]
+    public abstract class CrucibleQuestRequirementConfig : CruciblePackageConfigNode
```
This got a bit absurd but it ended up working pretty well. If you check out https://github.com/AndrewFahlgren/potioncraft-crucible/wiki/Crucible-Package-Reference#quest-requirements you can see it got a little complicated to explain. I am not sure if there is a better way to do this which may be more strait forward to the modder to implement.

RoboPhred replied:
The yaml api this creates looks pretty spot on to me, I think it's fine.

AndrewFahlgren committed:
Cleared trader inventory when specifying new items in the trader config

RoboPhred commented on Crucible.GameAPI/CrucibleQuest.cs:
```diff
@@ -0,0 +1,208 @@
+        /// <summary>
+        /// Gets or sets the minimum and maximum chapter needed to encounter this quest.
+        /// </summary>
+        public (int, int) MinMaxChapters
```
Lets avoid tupple return types and just make 2 properties for MinChapter and MaxChapter.  Tupple work is a bit advanced for the skill set that this API is aimed at, and I am worried people wont know how to set just one.

RoboPhred replied:
Alternatively, keep the tupple property, but add MinChapter and MaxChapter properties alongside it.

AndrewFahlgren replied:
Added two properties for min/max.

RoboPhred commented on Crucible.NPCs/Trader/CrucibleInventoryItemSoldByNpcStaticConfig.cs:
```diff
@@ -0,0 +1,55 @@
+        public void OnApplyConfiguration(CrucibleTraderNpcTemplate subject)
+        {
+            var inventoryItem = CrucibleInventoryItem.GetByName(this.InventoryItemName);
+            inventoryItem.ClosenessRequirement = this.ClosenessRequirement;
```
This seems to set a closeness requirement that is set to the item globally regardless of trader.  Is this the case?

If so, we should not allow a per-trader setting if different traders will simply override the global value for all traders.

AndrewFahlgren replied:
The closeness requirement actually is on a trader by trader basis. I left in the old closeness configuration option for ingredients which now acts a default closeness requirement. A closeness requirement can also be specified for each sold by configuration which will override the default if set.

RoboPhred commented on Crucible.GameAPI/CrucibleTraderNpcTemplate.cs:
```diff
@@ -47,6 +57,153 @@ internal CrucibleTraderNpcTemplate(NpcTemplate npcTemplate)
+        /// <summary>
+        /// Gets or sets the minimum and maximum karma range that this trader will spawn in.
+        /// </summary>
+        public (int, int) KarmaForSpawn
```
Another tupple that should be broken out int a max and min prop

AndrewFahlgren replied:
Added two properties for min/max.

AndrewFahlgren committed:
Added properties for min/max values individually for tupple values.
Added names to tuple values to clarify what Item1 and Item2 mean.
Refactored ingredients to allow for a default closeness requirement for an ingrendient and/or specify a closeness requirement for each individual trader

AndrewFahlgren committed:
Added examples directory and an example mod to Crucible.NPCs

AndrewFahlgren commented:
I also added an examples directory to Crucible.NPC with a mod I made while testing to serve as an example to modders. Is there a better place for this?

AndrewFahlgren committed:
Updated example mod to have appearance specified for trickster

AndrewFahlgren committed:
Added additional options to npc appearance to fix yaml errors

AndrewFahlgren committed:
Added an additional missing sprite

AndrewFahlgren committed:
Fixed compile error
Fixed issue where appearances parts were not named in the expected way by Potion Craft
Added additional null checks to appearance apply code

AndrewFahlgren commented:
Found a couple errors while testing NPC appearances more thoroughly. Updated PR with fixes and some updates to the test mod to include appearance information.
